### PR TITLE
Downsample v0.3

### PR DESCRIPTION
**Pull request by fangliu117**
Please review the refactored downsample_cells function at your convenience. Thanks!
Please find the relevant JIRA ticket and updated explanation of stratification at
https://tracker.nci.nih.gov/browse/DSP-186

### CODE REVIEW

File: src/spac/data_utils.py [hunk 1]
```diff
@@ -472,18 +471,18 @@ def select_values(data, annotation, values=None):
...
 
     Parameters
     ----------
-    data : pd.DataFrame
+    input_data : pd.DataFrame
         The input data frame.
-    annotation : str or list of str
+    annotations : str or list of str
         The column name(s) to downsample on. If multiple column names are
         provided, they are combined into a single column for downsampling.
```
Comment (georgezakinih) [inline src/spac/data_utils.py:487]: @fangliu117 , would you please elaborate how the two columns are combined in one? You mean the unique combinations of values in the two columns is generated?

File: src/spac/data_utils.py [hunk 2]
```diff
@@ -497,45 +496,45 @@ def downsample_cells(data, annotation, n_samples=None,
...
+    # Convert annotations to list if it's a string
+    if isinstance(annotations, str):
+        annotations = [annotations]
 
-    # Check if the columns to downsample on exists
-    for col in annotation:
-        if col not in data.columns:
+    # Check if the columns to downsample on exist
+    for col in annotations:
+        if col not in input_data.columns:
```
Comment (georgezakinih) [inline src/spac/data_utils.py:520]: @fangliu117 , can we use the util function to check for annotations?

File: src/spac/data_utils.py [hunk 3]
```diff
@@ -560,20 +559,22 @@ def downsample_cells(data, annotation, n_samples=None,
...
-            data = data.groupby(grouping_col).apply(
+            output_data = input_data.groupby(grouping_col).apply(
                 lambda x: x.head(min(n_samples, len(x)))
             ).reset_index(drop=True)
 
         # Remove the combined column if it exists
-        if len(annotation) > 1:
-            data = data.drop(columns=combined_col_name)
+        if len(annotations) > 1:
+            output_data = output_data.drop(columns=combined_col_name)
```
Comment (georgezakinih) [inline src/spac/data_utils.py:570]: @fangliu117 , I blelieve we can keep the combined col in the output data. It would be useful how the stratification took place.

File: src/spac/data_utils.py [hunk 4]
```diff
@@ -502,44 +503,73 @@ def downsample_cells(data, annotation, n_samples=None,
...
         # Stratify selection
         if stratify:
             # Determine frequencies of each group
-            freqs = data[annotation].value_counts(normalize=True)
+            freqs = data[grouping_col].value_counts(normalize=True)
             n_samples_per_group = (freqs * n_samples).astype(int)
+
+            # Increase the number of samples for groups with 0
+            # calculated samples.
+            n_samples_per_group = n_samples_per_group.apply(
```
Comment (georgezakinih) [inline src/spac/data_utils.py:542]: @fangliu117 , would you add a unit test for this case (group that has zero samples?)

File: tests/test_data_utils/test_downsample_cells.py [hunk 5]
```diff
@@ -1,98 +1,150 @@
...
+                'r1', 'r1', 'r1', 'r2', 'r2', 'r1', 'r1', 'r2', 'r3', 'r4'
+            ],
+            'value': list(range(1, 11))  # Simple values from 1 to 10
         })
-        self.annotation = 'annotation'
+
+        self.annotation_single = 'annotation'
+        self.annotation_multi = ['animal', 'region']
 
     def test_annotation_existence(self):
```
Comment (georgezakinih) [inline tests/test_data_utils/test_downsample_cells.py]: @fangliu117 , if you use the util function for checking annotation, we can remove this unit test.

File: tests/test_data_utils/test_downsample_cells.py [hunk 6]
```diff
@@ -1,98 +1,150 @@
...
+                self.df_multi_random,
+                ['nonexistent_column', 'region']
+            )
 
     def test_downsample_without_stratify(self):
         """Downsample without stratify"""
         n_samples = 20
-        df_downsampled = downsample_cells(
-            self.df, self.annotation,
+        df_downsampled_single = downsample_cells(
```
Comment (georgezakinih) [inline tests/test_data_utils/test_downsample_cells.py:66]: @fangliu117 , let's split this unit test into two: one for downsample single, and one for downsample multi. This way every test would be concise for a given behavior in the code.

File: tests/test_data_utils/test_downsample_cells.py [hunk 7]
```diff
@@ -1,98 +1,150 @@
...
 
-    def test_stratification_frequency(self):
-        """Test the frequency of stratification"""
-        n_samples = 10
+        actual_counts = downsampled_df[combined_col].value_counts()
+
+        for key, expected_count in expected_counts.items():
+            self.assertEqual(expected_count, actual_counts.get(key, 0))
+
+        # Remove the combined column after checks
```
Comment (georgezakinih) [inline tests/test_data_utils/test_downsample_cells.py:105]: @fangliu117, I am not sure why this code is in the unit test?

File: tests/test_data_utils/test_downsample_cells.py [hunk 8]
```diff
@@ -1,98 +1,150 @@
...
+        # Ensure the combined column exists in the downsampled_df
+        if combined_col not in downsampled_df.columns:
+            downsampled_df[combined_col] = downsampled_df[columns].apply(
+                lambda row: '_'.join(row.values.astype(str)), axis=1
+            )
 
-    def test_stratification_frequency(self):
-        """Test the frequency of stratification"""
-        n_samples = 10
+        actual_counts = downsampled_df[combined_col].value_counts()
```
Comment (georgezakinih) [inline tests/test_data_utils/test_downsample_cells.py:100]: @fangliu117, in the unit test, calucate the "actual_counts" manually and put it as a constant in the code. Here you run the same code that put in the downsample function.

File: tests/test_data_utils/test_downsample_cells.py [hunk 9]
```diff
@@ -1,98 +1,150 @@
...
+            # Only drop if it's a multi annotation
+            downsampled_df.drop(columns=combined_col, inplace=True)
 
-        # Downsampling with stratifying
+    def test_stratification_frequency_single_obs(self):
+        """Test the frequency of stratification for single annotation."""
+        n_samples = 10
         df_downsampled_stratified = downsample_cells(
-            self.df, self.annotation,
+            self.df_single_fixed, self.annotation_single,
```
Comment (georgezakinih) [inline tests/test_data_utils/test_downsample_cells.py]: @fangliu117, would you like to use df_single_random instead which has 100 cells? This way we can see that downsampling actually took place. you can keep this test as well as a corner case.

File: tests/test_data_utils/test_downsample_cells.py [hunk 10]
```diff
@@ -1,98 +1,150 @@
...
-        annotation_value_counts = \
-            df_downsampled_stratified[annotation].value_counts()
-        self.assertEqual(
-            annotation_value_counts['annotation1'], 5,
-            'Number of samples in annotation1 does not match expected count'
-        )
-        self.assertEqual(
-            annotation_value_counts['annotation2'], 3,
-            'Number of samples in annotation2 does not match expected count'
+    def test_stratification_frequency_multi_obs(self):
```
Comment (georgezakinih) [inline tests/test_data_utils/test_downsample_cells.py:128]: @fangliu117 , would you like to add a test case where there is an actual downsampling (e.g., start with with larger number of cells).

File: src/spac/data_utils.py (hunk 1 above)
Comment (fangliu117) [reply src/spac/data_utils.py:487]: @georgezakinih Thank you for your question. When combining the two columns:
1.	The values of the specified annotation columns for each row are concatenated using an underscore. So if one column has the value "a" and the other "x" for a particular row, the combined column for that row would have the value "a_x".
2.	Yes, the unique combinations of values from the two columns are effectively generated. The combined column represents all possible combinations of values from the original columns. When we group by this combined column later in the code, we inherently work with these unique combinations, which allows us to treat each unique combination as a distinct group. For instance, rows with the same "a_x" combination will be grouped together.

File: src/spac/data_utils.py (hunk 2 above)
Comment (fangliu117) [reply src/spac/data_utils.py:520]: Yes, update it.

File: src/spac/data_utils.py (hunk 3 above)
Comment (fangliu117) [reply src/spac/data_utils.py:570]: Yes, remove these lines.

File: src/spac/data_utils.py (hunk 1 above)
Comment (fangliu117) [reply src/spac/data_utils.py]: Update: to make it clearer, I have updated the docstring with more information in the new commit https://github.com/FNLCR-DMAP/SCSAWorkflow/pull/102/commits/f53cf81de01872afbe0488d7e44588f89aa27b76

File: src/spac/data_utils.py (hunk 2 above)
Comment (fangliu117) [reply src/spac/data_utils.py]: Update: I tried to use util function, which requires the Anndata as input, while for most of functions of data_utils, we have inputs as dataframe. Since this function is for data preprocessing, and to keep the consistence, I leave these lines now, but if needed, I'd like do more modification.

File: src/spac/data_utils.py (hunk 4 above)
Comment (fangliu117) [reply src/spac/data_utils.py]: Yes, added.

File: tests/test_data_utils/test_downsample_cells.py (hunk 5 above)
Comment (fangliu117) [reply tests/test_data_utils/test_downsample_cells.py]: Since util function didn't apply, currently, I keep it here, but will be happy to change if needed.

File: tests/test_data_utils/test_downsample_cells.py (hunk 6 above)
Comment (fangliu117) [reply tests/test_data_utils/test_downsample_cells.py:43]: Yes, split them.

File: tests/test_data_utils/test_downsample_cells.py (hunk 7 above)
Comment (fangliu117) [reply tests/test_data_utils/test_downsample_cells.py]: Actually this was a helper function to define the expect counts. However, in the new commit, we include the manually counted true numbers of expected count, therefore this function was removed.

File: tests/test_data_utils/test_downsample_cells.py (hunk 9 above)
Comment (fangliu117) [reply tests/test_data_utils/test_downsample_cells.py]: yes, df_single_random was applied.

File: tests/test_data_utils/test_downsample_cells.py (hunk 10 above)
Comment (fangliu117) [reply tests/test_data_utils/test_downsample_cells.py]: yes, a large dataset for testing multiple annotation was applied.

File: tests/test_data_utils/test_downsample_cells.py (hunk 8 above)
Comment (fangliu117) [reply tests/test_data_utils/test_downsample_cells.py]: yes, changed it.

File: src/spac/data_utils.py (hunk 1 above)
Comment (georgezakinih) [reply src/spac/data_utils.py]: Thanks @fangliu117, yes that's a clever idea.

File: src/spac/data_utils.py (hunk 2 above)
Comment (georgezakinih) [reply src/spac/data_utils.py]: Thanks @fangliu117, yes your comments make sense. You don't need to update these lines.

File: tests/test_data_utils/test_downsample_cells.py [hunk 11]
```diff
@@ -60,89 +38,148 @@ def test_annotations_existence(self):
...
             n_samples=n_samples, stratify=True, rand=True
         )
+
+        # Expected counts based on stratified sampling
+        total_cells = len(self.df_single_random)
         expected_counts = {
-            'annotations1': 6,
-            'annotations2': 3,
-            'annotations3': 1
+            'annotations1': round(50/total_cells * n_samples),
```
[the remark this answers is not in the document]
Comment (fangliu117) [reply tests/test_data_utils/test_downsample_cells.py:130]: Thanks @georgezakinih for the suggestion. Yes, I did hardcode the GT and add the comment.

File: tests/test_data_utils/test_downsample_cells.py [hunk 12]
```diff
@@ -60,89 +38,148 @@ def test_annotations_existence(self):
...
+            'annotations': ['group_large'] * 990 + ['group_small'] * 10,
+            'value': list(range(1, 1001))
+        })
+        n_samples = 20
+        df_downsampled = downsample_cells(
+            df_zero_sample, 'annotations',
+            n_samples=n_samples, stratify=True
+        )
+        actual_counts = df_downsampled['annotations'].value_counts()
+        self.assertTrue(actual_counts.get('group_small', 0) >= 1)
```
[the remark this answers is not in the document]
Comment (fangliu117) [reply tests/test_data_utils/test_downsample_cells.py]: Thanks, yes, I changed it.

File: src/spac/data_utils.py [hunk 13]
```diff
@@ -471,105 +471,145 @@ def select_values(data, annotation, values=None):
...
     """
-    Reduces the number of rows in the input_data by either selecting n_samples
-    from  every possible value of annotations, or returning n_samples
-    stratified by the frequency of values in annotations.
+    Custom downsampling of data based on one or more annotations.
+
+    This function offers two primary modes of operation:
+    1. **Grouping (stratify=False)**:
+       - For a single annotation: The data is grouped by unique values of the
+         annotation,and 'n_samples' rows are selected from each group.
```
[the remark this answers is not in the document]
Comment (fangliu117) [reply src/spac/data_utils.py]: Thanks, I corrected it.

File: src/spac/data_utils.py [hunk 14]
```diff
@@ -507,7 +508,7 @@ def downsample_cells(input_data, annotations, n_samples=None, stratify=False,
         - stratify=False: Returns 'n_samples' for each unique value (or
           combination) of annotations.
         - stratify=True: Returns a total of 'n_samples' stratified by the
-          frequency of annotations.
+          frequency of the every label or combined label in the annotation(s).
```
Comment (georgezakinih) [inline src/spac/data_utils.py:511]: @fangliu117 , please remove "the" form "of the every label" and ass 's' in "combined labels"

File: src/spac/data_utils.py [hunk 15]
```diff
@@ -576,17 +578,46 @@ def downsample_cells(input_data, annotations, n_samples=None, stratify=False,
 
         # Exclude groups with fewer samples than the min_threshold
         filtered_freqs = freqs[freqs * len(input_data) >= min_threshold]
+
+        # Log warning for groups that are excluded
+        excluded_groups = freqs[~freqs.index.isin(filtered_freqs.index)]
+        for group, count in excluded_groups.items():
+            logging.warning(
+                f"Group '{group}' with count {count} is excluded"
+                f" due to low frequency."
```
Comment (georgezakinih) [inline src/spac/data_utils.py:587]: @fangliu117, please also add a numeric value for the frequency.

File: src/spac/data_utils.py [hunk 16]
```diff
@@ -471,105 +471,145 @@ def select_values(data, annotation, values=None):
...
+
+        # Ensure each group has at least one sample if its frequency
+        # is non-zero
+        condition = samples_per_group == 0
+        samples_per_group[condition] = freqs[condition].apply(
+            lambda x: 1 if x > 0 else 0
+        )
+
+        # If have extra samples due to rounding, remove them from the
+        # largest groups
```
[the remark this answers is not in the document]
Comment (georgezakinih) [reply src/spac/data_utils.py:622]: @fangliu117, OK thanks for clarification. Let's for now put a warning showing how many cells were removed from the largest group.

[the remark this answers is not in the document]
Comment (fangliu117) [reply src/spac/data_utils.py:625]: Yes, I added.
